internal/dcache/rpc/client/client_pool.go

```diff
@@ -700,40 +976,55 @@ func (cp *clientPool) closeInactiveNodeClientPools() {
 
 // closeAllNodeClientPools closes all node client pools in the pool
 func (cp *clientPool) closeAllNodeClientPools() error {
-	// TODO: see if this is needed
-	cp.acquireLock()
-	defer cp.releaseLock()
+	// TODO: see if we need lock here, as we are closing all node client pools
+	log.Debug("clientPool::closeAllNodeClientPools: Closing all node client pools")
 
-	for key, ncPool := range cp.clients {
-		err := ncPool.closeRPCClients()
+	//
+	// Acquire write lock on the client pool to ensure that no other thread is accessing
+	// the client pool while we are closing all node client pools.
+	// This also waits till the read locks are released by the client pool operations
+	// like getRPCClient(), releaseRPCClient(), deleteAllRPCClients(), resetAllRPCClients(), etc.
+	//
+	cp.acquireRWMutexWriteLock()
+	defer cp.releaseRWMutexWriteLock()
```

Comment:
It'll be better to have closeAllNodeClientPools() wait for any allocated RPC clients to be released.
For that it should check that ncPool.numActive should be 0 before proceeding to close the connections from the channel. If not 0, it should sleep for a second and then try again. While sleeping release the write lock so that relaseRPCClient() can release clients.

Reply:
Done, updated